imxrt-hal/src/iomuxc/pin_config.rs

```diff
@@ -0,0 +1,197 @@
+    Disabled = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::RW::DSE_0_output_driver_disabled,
+
+    /// R0(150 Ohm @ 3.3V, 260 Ohm@1.8V)
+    R0 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::RW::DSE_1_R0_150_Ohm_3_3V_260_Ohm_1_8V,
+
+    /// R0/2
+    R0_DIV_2 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::RW::DSE_2_R0_2,
+
+    /// R0/3
+    R0_DIV_3 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::RW::DSE_3_R0_3,
+
+    /// R0/4
+    R0_DIV_4 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::RW::DSE_4_R0_4,
+
+    /// R0/5
+    R0_DIV_5 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::RW::DSE_5_R0_5,
+
+    /// R0/6
+    R0_DIV_6 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::RW::DSE_6_R0_6,
+
+    /// R0/7
+    R0_DIV_7 = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::DSE::RW::DSE_7_R0_7,
+}
+
+/// Speed Setting
+#[repr(u32)]
+pub enum Speed {
+    Speed0_50Mhz = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::SPEED::RW::SPEED_0_low_50MHz,
+    Speed1_100Mhz = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::SPEED::RW::SPEED_1_medium_100MHz,
+    //TODO fix svd files for Speed2, ref manual says 150MHz
+    Speed2_150Mhz = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::SPEED::RW::SPEED_2_medium_100MHz,
+    Speed3_200MHz = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::SPEED::RW::SPEED_3_max_200MHz,
+}
+
+/// Open Drain Setting
+#[repr(u32)]
+pub enum OpenDrain {
+    Disabled = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::ODE::RW::ODE_0_Open_Drain_Disabled,
+    Enabled = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::ODE::RW::ODE_1_Open_Drain_Enabled,
+}
+
+/// Pull Keeper Enable Setting
+#[repr(u32)]
+pub enum PullKeeperEnable {
+    Disabled = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::PKE::RW::PKE_0_Pull_Keeper_Disabled,
+    Enabled = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::PKE::RW::PKE_1_Pull_Keeper_Enabled,
+}
+
+/// Pull Keeper Select Setting
+#[repr(u32)]
+pub enum PullKeeperSelect {
+    Keeper = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::PUE::RW::PUE_0_Keeper,
+    Pull = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::PUE::RW::PUE_1_Pull,
+}
+
+/// Pull Up/Down Setting
+#[repr(u32)]
+pub enum PullUpDown {
+    PullDown_100KOhm = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::PUS::RW::PUS_0_100K_Ohm_Pull_Down,
+    PullUp_47KOhm = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::PUS::RW::PUS_1_47K_Ohm_Pull_Up,
+    PullUp_100KOhm = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::PUS::RW::PUS_2_100K_Ohm_Pull_Up,
+    PullUp_22KOhm = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::PUS::RW::PUS_3_22K_Ohm_Pull_Up,
+}
+
+/// Hysteresis Setting
+#[repr(u32)]
+pub enum Hysterisis {
+    Disabled = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::HYS::RW::HYS_0_Hysteresis_Disabled,
+    Enabled = ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::HYS::RW::HYS_1_Hysteresis_Enabled,
+}
+
+/// Pin Config
+/// A Builder pattern struct to generate a PinConfig that can then be used by
+/// the Pad to adjust the PAD_CTL_PAD register in as few operations as possible
+pub struct PinConfig {
+    pub mask: u32,
+    pub value: u32,
```

Comment:
I was debating that myself, the thing that made me put it pub was the idea that I could, in for example the imxrt1060evk-bsp, define the led function with a const PinConfig value using the macro. I don't believe it'd be possible if this were private.

Reply:
If we want the best of `const` while maintaining member privacy, could we support a `const fn` as the constructor?

```rust
pub const fn config(
    slew_rate: SlewRate,
    drive_strength: DriveStrength,
    open_drain: OpenDrain,
    /* other fields */
) -> PinConfig {
    /* ... */
}
```

The arguments are enums that could be evaluated at compile time. They're implemented by combining the existing RAL values with the RAL offsets.

```rust
use ral::iomuxc::SW_PAD_CTL_PAD_GPIO_EMC_00::*;

#[repr(u32)]
pub enum SlewRate {
    Slow = SRE::RW::SRE_0_Slow_Slew_Rate << SRE::RW::offset,
    Fast = SRE::RW::SRE_1_Fast_Slew_Rate << SRE::RW::offset,
}
```

The implementation of `config` just needs to OR the enums together into a `u32`. If we require that users always specify each field (or support a const `default()` function for convenience), we might not need to support the `is_modify()` API. It always becomes a store in the `configure()` method, since users are required to think though every field's values.

Reply:
I wasn't sure to be honest on the const fn constructors, I was under the impression const fn wasn't stabilized yet. If this works, I think its a great idea.

Reply:
I added the const constructor as described here as PinConfig::with_all()

There's also a PinConfig::with_none() which allows incrementally constructing a value using the const fn constructor funcs you described elsewhere. I'm interested to find out how smart const fn is at removing baggage at compile time, though even if it exists still I don't believe it will be very costly for an infrequently changed value.